lib/howl/ui/command_line.moon

```diff
@@ -0,0 +1,503 @@
+-- Copyright 2012-2015 The Howl Developers
+-- License: MIT (see LICENSE.md at the top-level directory of the distribution)
+
+Gtk = require 'ljglibs.gtk'
+import bindings, dispatch, interact, signal from howl
+import Matcher from howl.util
+import PropertyObject from howl.aux.moon
+import TextWidget, NotificationWidget, IndicatorBar, StyledText from howl.ui
+import highlight, markup, style, theme from howl.ui
+
+-- used to generate a unique id for every new activity
+id_counter = 0
+
+class CommandLine extends PropertyObject
+  new: (@window) =>
+    super!
+    @bin = Gtk.Box Gtk.ORIENTATION_HORIZONTAL
+    @box = nil
+    @command_widget = nil
+    @notification_widget = nil
+    @header = nil
+    @indic_title = nil
+    @showing = false
+    @spillover = nil
+    @running = {}
+    @_command_history = {}
+    @aborted = {}
+    @next_run_queue = {}
+    @auto_submit_once = false
+
+  @property current: get: => @running[#@running]
+
+  @property stack_depth: get: => #@running
+
+  @property command_history: get: => moon.copy @_command_history
+
+  run_auto_submit: (...) =>
+    @auto_submit_once = true
+    @run ...
+
+  run: (activity_spec, ...) =>
```

Comment:
This is an overly long method. After perusing it I can't find any obvious candidates for extraction, but perhaps you can?